Make container runtime message types stricter

noencke opened the pull request:
This PR seeks to make the types for messages sent to and from the container runtime more explicit. Messages received have a slightly different signature than messages sent, in order to support back and forward compatibility. Messages now use a discriminated union over their type so that their contents is strongly typed rather than being `any`.

Behavior changes:
- `unpack` in `remoteMessageProcessor` only adds `compatDetails` property if it exists in inner contents
- `parseLocalOpContent` in `containerRuntime` returns all contents that were serialized rather than limiting to known three: `type`, `contents`, and `compatDetails`.

noencke committed:
Add type union for container runtime messages. Minor cleanup.

noencke committed:
Merge remote-tracking branch 'upstream/main' into message-union

noencke committed:
Make unknown message part of local message

jason-ha commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
  * This way stringified values can be compared.
  */
-export interface ContainerRuntimeMessage {
+export interface ContainerRuntimeMessage<TType, TContents> {
```
>TType [](http://example.com/codeflow?start=41&length=5)

```suggestion
TType extends ContainerMessageType
```

jason-ha commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
+}
+
 /**
  * An unpacked ISequencedDocumentMessage with the inner ContainerRuntimeMessage type/contents/etc
```
>ContainerRuntimeMessage [](http://example.com/codeflow?start=56&length=23)

Needs updated and perhaps applied to Outbound type as well.

markfields commented on packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:
```diff
@@ -1037,8 +1038,8 @@ describe("Runtime", () => {
-				const futureRuntimeMessage: ContainerRuntimeMessage = {
-					type: "FROM_THE_FUTURE" as ContainerMessageType,
+				const futureRuntimeMessage: UnknownContainerRuntimeMessage = {
+					type: "__unknown__",
```
I liked Jason's point that `__unknown__` is _never_ going to be the value and could be misleading to do so in tests.  Better to do `"FROM_THE_FUTURE" as unknown as "__unknown__"`.

noencke replied:
Presumably, these tests also want to use a value that is never going to be legitimate (because, if such a value were introduced, the test will no longer be testing "the future" and will be merely testing the present). So, I don't actually see any difference between literally using `__unknown__` and using something that casts to `__unknown__`, except that the latter is longer.

jason-ha replied:
I would explicitly use something different. I don't want to see "__unknown__" anywhere else in the codebase. I would even add a linter rule for it. Perhaps "__unknown__" should be even more verbose "__unknown_ContainerRuntimeMessage_type__" or such. Can make is as stupid long as we want because it will never be an actual value.

markfields replied:
Up to you, my theoretical concern is that someone would look at this and thing the _value_ `__unknown__` means something, but hopefully the underscores would tip them off or something. We're using the _type_ `__unknown__` in place of `Exclude<string, "x" | "y" | "z">` since that's not supported, so any value is a good example.

noencke replied:
Alright, you've convinced me.

jason-ha commented:
> 			case ContainerMessageType.IdAllocation:

Can do away with these casts to by using `opContents.`.
Line 2126 should probably go away as there are then single uses of `type` and `compatDetails`.

---
Refers to: packages/runtime/container-runtime/src/containerRuntime.ts:2132 in 22d20d1. [](commit_id = 22d20d1e90f398a5ff243d400495e368bc4a9448, deletion_comment = False)

markfields commented on api-report/container-runtime.api.md:
```diff
@@ -240,14 +245,31 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents
+// Warning: (ae-forgotten-export) The symbol "IDataStoreAliasMessage" needs to be exported by the entry point index.d.ts
+//
+// @public (undocumented)
+export type ContainerRuntimeAliasMessage = ContainerRuntimeMessage<ContainerMessageType.Alias, IDataStoreAliasMessage>;
+
+// @public (undocumented)
+export type ContainerRuntimeBlobAttachMessage = ContainerRuntimeMessage<ContainerMessageType.BlobAttach, undefined>;
+
+// @public (undocumented)
+export type ContainerRuntimeChunkedOpMessage = ContainerRuntimeMessage<ContainerMessageType.ChunkedOp, IChunkedOp>;
+
+// @public (undocumented)
+export type ContainerRuntimeDataStoreOpMessage = ContainerRuntimeMessage<ContainerMessageType.FluidDataStoreOp, IEnvelope>;
```
Do we need to export these?  Probably ties down to types exported below.  Would be good to take a pass and see if we can _stop_ exporting as much as possible here, even `ContainerRuntimeMessage` - not sure why it's there.

noencke replied:
Good question. I assumed that we do but if we can not export any of it that'd be awesome. I'll try it and see what happens.

markfields replied:
Cool.  I think you'll have to deprecate `ContainerRuntimeMessage` as-is and give the generic one a new name, right?

noencke replied:
Hmm, I suppose so. Now that you point it out, I'm actually surprised we can make the change as it currently is - adding required generic parameters - without breaking back-compat tests.

noencke replied:
Looks like there's a tinylicious test that uses `ContainerRuntimeMessage` (sigh). But that's it, lol.

jason-ha replied:
@noencke , where is the tinylicious use? Search tells me there are no such uses.

noencke replied:
My bad, I meant `ContainerMessageType`, not `ContainerRuntimeMessage`. I reduced the exports to the minimal without having to mess with back-compat tests.

jason-ha commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
+	contents: TContents;
 	/** Info describing how to handle this op in case the type is unrecognized (default: fail to process) */
 	compatDetails?: IContainerRuntimeMessageCompatDetails;
 }
```
This is from the past, but is it expected to exist or not exist on any of these specific messages?
(I am have the impression that it is a future proofing aspect but not expected for the already known cases. Haven't read the code.)

markfields replied:
That's correct, no current code puts `compatDetails` on.  In coming sprints I'll add a case that does though.  Down the road I'll probably remove it once enough time / versions have passed.

jason-ha replied:
Sure. So in that future thing you will do (correct me if I am wrong), there will be a new ContainerRuntimeMessageType entry. The outbound type must provide `compatDetails` and the inbound case could have it but at most it would be optional (since you want to remove it at some point).
So here and now, that should only appear on the Unknown case.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
 	compatDetails?: IContainerRuntimeMessageCompatDetails;
 }
 
+export type ContainerRuntimeDataStoreOpMessage = ContainerRuntimeMessage<
```
Let's put all these types (and the existing related types/code) in a separate file, either in this PR or I can in a follow-up.

noencke replied:
Good idea, I can do it in this PR.

noencke replied:
They're now in `messageTypes.ts`

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
+	| ContainerRuntimeRejoinMessage
+	| ContainerRuntimeAliasMessage
+	| InTransitContainerRuntimeIdAllocationMessage
+	| UnknownContainerRuntimeMessage;
```
```suggestion
	// Inbound messages may include unknown types from other clients, so we include that as a special case here
	| UnknownContainerRuntimeMessage;
```

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
+	| ContainerRuntimeRejoinMessage
+	| ContainerRuntimeAliasMessage
+	| LocalContainerRuntimeIdAllocationMessage
+	| UnknownContainerRuntimeMessage;
```
```suggestion
	// In rare cases (e.g. related to stashed ops) we could have a local copy of a message with an unknown type
	| UnknownContainerRuntimeMessage;
```

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
+	| LocalContainerRuntimeIdAllocationMessage
+	| UnknownContainerRuntimeMessage;
+
+/** A {@link ContainerRuntimeMessage} that is being sent to the server from the container runtime. */
```
With the three similar-but-different types here, is there some kind of test we could write that would fail to build if these types diverged?  Like, each of these types must cover all the known `ContainerMessageType` values - so having a switch statement in a test file or something?  Just thinking out loud here.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -242,33 +249,124 @@ export interface IContainerRuntimeMessageCompatDetails {
+	ContainerMessageType.IdAllocation,
+	IdCreationRangeWithStashedState
+>;
+export type InTransitContainerRuntimeIdAllocationMessage = ContainerRuntimeMessage<
```
I was thinking about this, maybe we can drop the `InTransit` from the name - the type over the wire can be the "canonical" type, and then the `Local`-named version has extra.  Or if you like it leave it, that phrase "in transit" was my idea but I never really like it :P

jason-ha replied:
FWIW I am good with dropping InTransit- and just keeping Local-.

markfields commented on packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:
```diff
@@ -1053,8 +1054,8 @@ describe("Runtime", () => {
 			it("process remote op with unrecognized type and 'FailToProcess' compat behavior", async () => {
-				const futureRuntimeMessage: ContainerRuntimeMessage = {
-					type: "FROM_THE_FUTURE" as ContainerMessageType,
+				const futureRuntimeMessage: UnknownContainerRuntimeMessage = {
```
I love this typing!

jason-ha replied:
but we don't want that type really doing out; so, it is back to being more or less free for all which is what the test is doing.

jason-ha commented on packages/runtime/container-runtime/src/index.ts:
```diff
@@ -5,7 +5,23 @@
+	LocalContainerRuntimeIdAllocationMessage,
+	OutboundContainerRuntimeAttachMessage,
+	OutboundSequencedContainerRuntimeMessage,
 	IContainerRuntimeMessageCompatDetails,
```
Looks like I need to introduce this project to `type` imports including an eslint plugin that will fix and enforce them.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2015,18 +2113,18 @@ export class ContainerRuntime
 		// Need to parse from string for back-compat
-		const { type, contents, compatDetails } = this.parseOpContent(op);
-		switch (type) {
+		const opContents = this.parseLocalOpContent(serializedOpContent);
```
For follow-up: I looked around the code a bit and in the cases I checked, `serializedOpContent` appears like it'll be serialized `Outbound` not `Local` message contents.

cc @jason-ha

markfields replied:
Actually I think we can make this change now, updating `parseLocalOpContent` to be `parseOutboundOpContent` with the corresponding type.

I'm trying to remember how we concluded this should be `Local`.  Maybe the ID Compressor handling was incorrect and we worked back from there...?

noencke replied:
Right - the IdAllocation case expects there to be stashed state on the op here. I would expect a stashed op that's being applied to be "local" not "outgoing". I guess the problem is that the op _was_ outgoing when it was sent, but now it is local? So, does the distinction between local and outgoing even make sense?

markfields replied:
Let's discuss after you merge this, as-is more closely matches the code being touched, even if I anticipate it will clash elsewhere in the next wave of changes.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2015,18 +2113,18 @@ export class ContainerRuntime
-			JSON.parse(serializedContent);
-		assert(type !== undefined, 0x6d6 /* incorrect op content format */);
-		return { type, contents, compatDetails };
+	private parseLocalOpContent(serializedContents?: string): LocalContainerRuntimeMessage {
```
Should be `Outbound` not `Local`, I'm pretty sure.  See comment thread below at the top of `applyStashedOp`

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2036,7 +2134,7 @@ export class ContainerRuntime
 					0x67b /* IdCompressor should be defined if enabled */,
 				);
-				return this.applyStashedIdAllocationOp(contents as IdCreationRangeWithStashedState);
+				return this.applyStashedIdAllocationOp(opContents.contents);
```
I wonder how many `as`'s we can kick out this week 😇

jason-ha replied:
That is the metric, right?

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2158,7 +2256,15 @@ export class ContainerRuntime
-			this.processCore(message, local, modernRuntimeMessage);
+			if (modernRuntimeMessage) {
+				this.processCore({
+					message: message as InboundSequencedContainerRuntimeMessage,
```
```suggestion
					// Cast it since we expect it to be this based on modernRuntimeMessage computation above
					message: message as InboundSequencedContainerRuntimeMessage,
```

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2171,10 +2277,19 @@ export class ContainerRuntime
 	 * @param modernRuntimeMessage - Does this appear like a current ContainerRuntimeMessage?
```
Update these `@param` comments?  Not sure what the best way to document these are, can you do `@param args.message` or something...?

noencke replied:
The only way I know of getting docs that also provide mouse-hover docs is to extract the type of `args` into an interface and doc each param on the interface. Since this is an internal method, I'd rather not, and instead just do `@param args.message`, etc.

jason-ha replied:
Created helper type gave it tsdoc comment

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2015,18 +2113,18 @@ export class ContainerRuntime
-		const { type, contents, compatDetails } = this.parseOpContent(op);
-		switch (type) {
+		const opContents = this.parseLocalOpContent(serializedOpContent);
+		const { type, contents, compatDetails } = opContents;
```
Delete this and inline `opContents.` everywhere. It's more consistent and gets rid of several `as`'s

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2228,10 +2345,10 @@ export class ContainerRuntime
 		local: boolean,
 		expectRuntimeMessageType: boolean,
-	): asserts message is SequencedContainerRuntimeMessage {
+	): asserts message is InboundSequencedContainerRuntimeMessage {
```
Just realized this isn't true - if it's a system op this will return early from the default case and this type assertion doesn't hold.
```suggestion
	): void {
```

markfields replied:
ACTUALLY!  Now that we have the `Inbound` type which includes the `__unknown__` case it seems fair to just type `message: InboundContainerRuntimeMessage` in the function signature.  TS seems to allow it even though at the callsite `message.type` could be any `string`... Not sure why but it works for our case.  This function has to handle the unknown cases.

jason-ha replied:
I was thinking similarly.

In all cases (see below):
1. The optimistic destructure should go away.
2. The return should be void.

To simply type message as InboundSequencedContainerRuntimeMessage however, the call-site will require `message as InboundSequencedContainerRuntimeMessage`.
*Almost* alternatively this could accept `InboundSequencedContainerRuntimeMessage | ISequencedDocumentMessage` and the access to `compatDetails` would need guarded by (`'compatDetails' in message`. But then message type going into `compatBehaviorAllowsMessageType` might still be string. Perhaps sync up in the morning.

jason-ha replied:
I was able to leverage the callers structure to get to `InboundSequencedContainerRuntimeMessage | ISequencedDocumentMessage` or `InboundSequencedContainerRuntimeMessage` when modern and that let some `as` go away, but it isn't good enough as your `any` comment noted.

jason-ha commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2158,7 +2256,15 @@ export class ContainerRuntime
+					local,
+					modernRuntimeMessage,
+				});
+			} else {
```
I don't think we fix it in this PR, but one thing that has been bugging me is that there isn't a strict assertion/typing/rule that the message.type for all of the messages that come out of remoteMessageProcessor.process must be MessageType.Operation or not just as messageArg.type is. Seems fragile.

markfields replied:
If I understand what you're saying, it doesn't actually hold - Rather, if `messageArg.type` is `MessageType.Operation`, then the output will be "unpacked" such that each `message.type` will be `ContainerMessageType` (or unknown in the future case).  Otherwise, _I believe_ the output will be a singleton array containing `messageCopy`.

markfields replied:
And yeah, I'd like to confirm and codify that as well.

jason-ha replied:
Adding comment.

noencke committed:
Move types to seperate file and minimize exports

markfields commented on packages/runtime/container-runtime/src/messageTypes.ts:
```diff
@@ -0,0 +1,188 @@
+	| ContainerRuntimeAliasMessage
+	| InTransitContainerRuntimeIdAllocationMessage;
+
+/**
```
Reminder to fix this comment and put it on the Outbound one too

markfields commented on packages/runtime/container-runtime/src/index.ts:
```diff
@@ -28,6 +24,13 @@ export {
+export {
+	ContainerMessageType,
+	ContainerRuntimeMessage,
+	UnknownContainerRuntimeMessage,
```
Why?

markfields commented on packages/runtime/container-runtime/src/pendingStateManager.ts:
```diff
@@ -242,7 +242,9 @@ export class PendingStateManager implements IDisposable {
 	 */
-	public processPendingLocalMessage(message: SequencedContainerRuntimeMessage): unknown {
+	public processPendingLocalMessage(
+		message: ISequencedDocumentMessage & TypedContainerRuntimeMessage<string, unknown>,
```
Do we want an alias `AnyContainerRuntimeMessage` for this?

jason-ha replied:
There is an effective Any* used with the new helper `buildPendingMessageContent`. Uses of such things should be limited.

markfields commented on packages/runtime/container-runtime/src/test/types/validateContainerRuntimePrevious.generated.ts:
```diff
@@ -64,26 +64,26 @@ use_old_VariableDeclaration_AllowTombstoneRequestHeaderKey(
 declare function get_old_TypeAliasDeclaration_CompatModeBehavior():
     TypeOnly<old.CompatModeBehavior>;
-declare function use_current_TypeAliasDeclaration_CompatModeBehavior(
+declare function use_current_RemovedTypeAliasDeclaration_CompatModeBehavior(
```
Yeah since I'm 99.99% sure no one is using this it seems fine to just remove.  Typically we should deprecate though.

markfields commented on packages/runtime/container-runtime/src/test/types/validateContainerRuntimePrevious.generated.ts:
```diff
@@ -112,26 +112,26 @@ use_old_EnumDeclaration_CompressionAlgorithms(
 * Validate forward compat by using old type in place of current type
 * If breaking change required, add in package.json under typeValidation.broken:
-* "EnumDeclaration_ContainerMessageType": {"forwardCompat": false}
+* "RemovedEnumDeclaration_ContainerMessageType": {"forwardCompat": false}
```
This I would deprecate and leave in `index` for now.  We will have to deal with the strikethrough in our own codebase for a few months.

markfields commented on packages/runtime/container-runtime/src/test/types/validateContainerRuntimePrevious.generated.ts:
```diff
@@ -160,26 +160,26 @@ use_old_ClassDeclaration_ContainerRuntime(
 * Validate forward compat by using old type in place of current type
 * If breaking change required, add in package.json under typeValidation.broken:
-* "InterfaceDeclaration_ContainerRuntimeMessage": {"forwardCompat": false}
+* "RemovedInterfaceDeclaration_ContainerRuntimeMessage": {"forwardCompat": false}
```
Leave in index.ts to avoid the breaking change.

markfields commented on packages/runtime/container-runtime/src/test/types/validateContainerRuntimePrevious.generated.ts:
```diff
@@ -496,26 +496,26 @@ use_old_InterfaceDeclaration_IConnectableRuntime(
 * Validate forward compat by using old type in place of current type
 * If breaking change required, add in package.json under typeValidation.broken:
-* "InterfaceDeclaration_IContainerRuntimeMessageCompatDetails": {"forwardCompat": false}
+* "RemovedInterfaceDeclaration_IContainerRuntimeMessageCompatDetails": {"forwardCompat": false}
```
This one's fine to cut without deprecating

markfields commented on packages/runtime/container-runtime/src/test/opLifecycle/opDecompressor.spec.ts:
```diff
@@ -19,7 +20,7 @@ interface ITestMessageContents {
 
 function generateCompressedBatchMessage(length: number): ISequencedDocumentMessage {
-	const batch: ContainerRuntimeMessage[] = [];
+	const batch: TypedContainerRuntimeMessage<ContainerMessageType.FluidDataStoreOp, string>[] = [];
```
Ummm what why isn't it `IEnvelope`  -_-

markfields replied:
I guess it just needs something and doesn't care about the actual message format.  Maybe just put a comment saying "this isn't what any real message looks like, it's just for testing" or something.

jason-ha replied:
Feels like TypedContainerRuntimeMessage shouldn't be exported from messageTypes if it can be helped.
Definitely need to call out things that aren't real.

markfields commented on packages/runtime/container-runtime/src/opLifecycle/remoteMessageProcessor.ts:
```diff
@@ -110,11 +110,11 @@ const copy = (remoteMessage: ISequencedDocumentMessage): ISequencedDocumentMessa
+): asserts message is InboundSequencedContainerRuntimeMessage {
+	const innerContents = message.contents as InboundContainerRuntimeMessage;
 
 	// We're going to turn message into a SequencedContainerRuntimeMessage in-place
```
nit
```suggestion
	// We're going to turn message into an InboundSequencedContainerRuntimeMessage in-place
```

noencke committed:
Cleanup, PR suggestions

markfields committed:
Make unpack function contents more readible (renaming only)

jason-ha committed:
refactor: type tightening
- remove compatDetails from current known messages
+ add RecentlyAddedContainerRuntimeMessageDetails with compatDetails
- remove UnknownRuntimeMessageType
+ fix some comment references

markfields committed:
Merge remote-tracking branch 'upstream/main' into pr/noencke/17285

markfields commented on packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:
```diff
@@ -1037,13 +1040,21 @@ describe("Runtime", () => {
 				};
 
-				const packedOp: Partial<ISequencedDocumentMessage> = {
+				const packedOp: Omit<
```
@jason-ha This doesn't feel worth it to me for tests.  I'm not as worried about playing fast and loose with types in a unit test since it's about driving the runtime a certain way and validating the results.  But I also don't mind if you do :)

jason-ha replied:
The part I liked was that I could see what things in contract the test was skipping. Ideally the processing wouldn't say those are needed if they aren't. Perhaps some refactor in future can better separate what matters when.

jason-ha committed:
commit generation changes

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2254,20 +2218,25 @@ export class ContainerRuntime
 					0x67c /* IdCompressor should be defined if enabled */,
 				);
-				this.idCompressor.finalizeCreationRange(message.contents as IdCreationRange);
+				this.idCompressor.finalizeCreationRange(message.contents);
```
My IDE is showing `message.contents: any` here.  I think this code should be operating under the assumption that the thing is a Runtime thing.  `|`ing in `ISequencedDocumentMessage` is watering this down I think

jason-ha replied:
I am seeing that too.
Even if I temporarily move the `!messageWithContext.modernRuntimeMessage` check to top to narrow to `InboundSequencedContainerRuntimeMessage` it doesn't get to just `ISequencedDocumentMessage & InTransitContainerRuntimeIdAllocationMessage`
(like it gets to `ISequencedDocumentMessage & UnknownContainerRuntimeMessage` for the default case).

We need `ISequencedDocumentMessage['contents']` to be `unknown` instead of `any` to get the narrowing. But even then that is only if the `modernRuntimeMessage` check is performed, which I think is a no go.

So we need something that better. Probably a type that omits `type` and/or at least  `contents` from `ISequencedDocumentMessage`. Thinking...

markfields replied:
I vote to leave it for now and try a scoped follow-up PR to improve it.

jason-ha replied:
The problem is that it isn't an improvement. We should at least restore the as.

markfields commented on packages/runtime/container-runtime/src/containerRuntime.ts:
```diff
@@ -2015,28 +1982,27 @@ export class ContainerRuntime
-		return { type, contents, compatDetails };
+	private parseLocalOpContent(serializedContents?: string): LocalContainerRuntimeMessage {
+		assert(serializedContents !== undefined, 0x6d5 /* content must be defined */);
+		const message: LocalContainerRuntimeMessage = JSON.parse(serializedContents);
```
This could be a behavior change if the object has more props than those 3.

markfields replied:
But it's "probably" fine based on my memory of callers, who pull the 3 pieces off it, and don't pass the object itself around.

jason-ha replied:
Per some discussion this probably better. If properties are added they would be dropped with prior logic.
The updated types make it hard to get full safety when handling properties one-by-one like before which is exactly true apart from just going with what you've got as we now do.

markfields reviewed (commented):
Ok I took one more pass to check for any behavior change, there is 1 but only in a hypothetical situation and it wouldn't matter (JSON.parse into a single variable instead of immediately destructuring the result to filter to only what we expect)

Ship it!

jason-ha committed:
test: use more correct op
+ remove TypedContainerRuntimeMessage helper export

jason-ha committed:
test: really fix 'em

jason-ha committed:
refactor: tighten InboundSequencedNonContainerRuntimeMessage
and add InboundSequencedNonContainerRuntimeMessageOrSystemMessage

jason-ha committed:
code review cleanups

markfields reviewed (approved):
🐿️ LGTM
